cryptoki/tests/basic.rs

```diff
@@ -2170,6 +2175,796 @@ fn ekdf_aes_cbc_encrypt_data() -> TestResult {
+        Attribute::Derive(true),
+    ];
+    let base_key = session.generate_key(&Mechanism::AesKeyGen, &base_template)?;
+
+    // The template of the key to derive
+    let derived_key_template = [
+        Attribute::Token(false),
+        Attribute::Private(false),
+        Attribute::Class(ObjectClass::SECRET_KEY),
+        Attribute::KeyType(KeyType::AES),
+        Attribute::ValueLen((AES256_BLOCK_SIZE as u64).into()),
+        Attribute::Encrypt(true),
+        Attribute::Decrypt(true),
+    ];
+
+    // Some variables we will use throughout
+    let counter_format = KbkdfCounterFormat::new(Endianness::Big, 16);
+    let dkm_length_format =
+        KbkdfDkmLengthFormat::new(KbkdfDkmLengthMethod::SumOfKeys, Endianness::Big, 16);
+
+    /* COUNTER-MODE */
+
+    // Instantiate KBKDF in counter-mode without additional keys
+    let data_params = [
+        PrfDataParam::new(PrfDataParamType::IterationVariable(Some(&counter_format))),
+        PrfDataParam::new(PrfDataParamType::ByteArray(b"test_label")),
+        PrfDataParam::new(PrfDataParamType::ByteArray(&[0])),
+        PrfDataParam::new(PrfDataParamType::ByteArray(b"\xfe\xed\xbe\xef")),
+        PrfDataParam::new(PrfDataParamType::DkmLength(&dkm_length_format)),
+    ];
+    let params = KbkdfParams::new(MechanismType::AES_CMAC, &data_params, None);
+
+    // Derive key
+    let derived_key_counter = session.derive_key(
+        &Mechanism::KbkdfCounter(params),
+        base_key,
+        &derived_key_template,
+    )?;
+
+    /* FEEDBACK-MODE - no IV */
+
+    // Instantiate KBKDF in feedback-mode without additional keys or an IV
+    let data_params = [
+        PrfDataParam::new(PrfDataParamType::IterationVariable(None)),
+        PrfDataParam::new(PrfDataParamType::Counter(&counter_format)),
+        PrfDataParam::new(PrfDataParamType::DkmLength(&dkm_length_format)),
+    ];
+    let params = KbkdfFeedbackParams::new(MechanismType::AES_CMAC, &data_params, None, None);
+
+    // Derive key
+    let derived_key_feedback_no_iv = session.derive_key(
+        &Mechanism::KbkdfFeedback(params),
+        base_key,
+        &derived_key_template,
+    )?;
+
+    /* FEEDBACK-MODE - with IV */
+
+    // Instantiate KBKDF in feedback-mode without additional keys but with an IV
+    let data_params = [
+        PrfDataParam::new(PrfDataParamType::IterationVariable(None)),
+        PrfDataParam::new(PrfDataParamType::Counter(&counter_format)),
+    ];
+    let params = KbkdfFeedbackParams::new(
+        MechanismType::AES_CMAC,
+        &data_params,
+        Some(b"some_initialization_vector"),
+        None,
+    );
+
+    // Derive key
+    let derived_key_feedback_iv = session.derive_key(
+        &Mechanism::KbkdfFeedback(params),
+        base_key,
+        &derived_key_template,
+    )?;
+
+    /* DOUBLE PIPELINE-MODE */
+
+    /* FIXME: NIST SP800-108 in double-pipeline mode is not yet supported by SoftHSM or Kryoptic */
```

Comment:
I think it'd be good to move all these unsupported cases into a separate test and mark them with `#[ignore = "unsupported by both softhsm and kryoptic"]`.

Reply:
Done!